File: BST/AVL/showlibia/Cpp-version/correct.cc
#include "avl.hh" 
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

struct Operation
{
    string operation;
    int value;
};

int load_operations(std::string_view filename, vector<Operation> &operations)
{
    ifstream file((string)filename);
    if (!file.is_open())
    {
        cerr << "Error opening file: " << filename << endl;
        return -1;
    }

    string line;
    while (getline(file, line))
    {
        istringstream iss(line);
        string op;
        int value;

        if (getline(iss, op, ',') && iss >> value)
        {
            operations.push_back({op, value});
        }
        else
        {
            cerr << "Warning: Malformed line: " << line << endl;
        }
    }

    file.close();
    return operations.size();
}

void process_operation(AVLTree &avl, const string &operation, int value, ofstream &output_file)
{
    if (operation == "insert")
    {
        avl.insert(value);
        output_file << "insert," << value << endl;
    }
    else if (operation == "search")
    {
        bool found = avl.search(value);
        output_file << "search," << value << "," << (found ? "true" : "false") << endl;
    }
    else if (operation == "delete")
    {
        bool found = avl.search(value);
        avl.remove(value);
        output_file << "delete," << value << (found ? "" : ",null") << endl;
    }
    else if (operation == "findmin")
    {
        AVLNode *min_node = avl.findMin();
        if (min_node)
        {
            output_file << "findmin," << min_node->data << endl;
        }
        else
        {
            output_file << "findmin,null" << endl;
        }
    }
    else if (operation == "findmax")
    {
        AVLNode *max_node = avl.findMax();
        if (max_node)
        {
            output_file << "findmax," << max_node->data << endl;
        }
        else
        {
            output_file << "findmax,null" << endl;
        }
    }
    else
    {
        cerr << "Warning: Unknown operation: " << operation << endl;
    }
}

void execute_operations(const vector<Operation> &operations, const string &output_filename)
{
    ofstream output_file(output_filename);
    if (!output_file.is_open())
    {
        cerr << "Error opening output file: " << output_filename << endl;
        return;
    }

    AVLTree avl;

    for (const auto &op : operations)
    {
        process_operation(avl, op.operation, op.value, output_file);
    }

    output_file.close();
}

int main()
{
    string input_filename = "../../../benchmark/benchmark_operations.csv";
    string output_filename = "correctness.csv";

    std::vector<Operation> operations;
    int count = load_operations(input_filename, operations);
    if (count < 0)
    {
        return 1;
    }

    execute_operations(operations, output_filename);

    return 0;
}

File: BST/AVL/showlibia/Go-version/correct.go
// correct.go
package main

import (
	"fmt"
	"log"
	"os"
)

func processOperationCorrect(root *AVLNode, operation string, value int, outputFile *os.File) *AVLNode {
	switch operation {
	case "insert":
		root = Insert(root, value)
		fmt.Fprintf(outputFile, "insert,%d\n", value)
	case "search":
		found := Search(root, value)
		fmt.Fprintf(outputFile, "search,%d,%t\n", value, found)
	case "delete":
		if Search(root, value) {
			root = Remove(root, value)
			fmt.Fprintf(outputFile, "delete,%d\n", value)
		} else {
			fmt.Fprintf(outputFile, "delete,%d,null\n", value)
		}
	case "findmin":
		minNode := FindMin(root)
		if minNode != nil {
			fmt.Fprintf(outputFile, "findmin,%d\n", minNode.Data)
		} else {
			fmt.Fprintf(outputFile, "findmin,null\n")
		}
	case "findmax":
		maxNode := FindMax(root)
		if maxNode != nil {
			fmt.Fprintf(outputFile, "findmax,%d\n", maxNode.Data)
		} else {
			fmt.Fprintf(outputFile, "findmax,null\n")
		}
	default:
		log.Printf("Warning: Unknown operation: %s\n", operation)
	}
	return root
}

func executeOperations(operations []Operation, outputFilename string) {
	outputFile, err := os.Create(outputFilename)
	if err != nil {
		log.Fatalf("Error opening output file: %v\n", err)
	}
	defer outputFile.Close()

	var root *AVLNode
	for _, op := range operations {
		root = processOperationCorrect(root, op.Operation, op.Value, outputFile)
	}
}

func correct() {
	inputFilename := "../../../benchmark/benchmark_operations.csv"
	outputFilename := "correctness.csv"

	operations, err := LoadOperations(inputFilename)
	if err != nil {
		log.Fatal(err)
	}

	executeOperations(operations, outputFilename)
}
